src/sparseml/transformers/compression/README.md

```diff
@@ -30,52 +30,131 @@ needed for decompression in the compressed state_dict:
 }
 ```
 
+Config information gets stored in the HF config file
+```json
+// config.json
+{
+    "sparsity_config": {
+        "format": "sparse_bitmask", // "dense_sparsity" for original tensor format
+
+        // informational
+        "sparsity_structure": "unstructured", // or 2:4, 8:16 etc...
+        "global_sparsity": "0.5"
+    }
+}
+```
+
+## Saving/Loading Interface 
+
+Loading in a compressed model requires no interface changes
+
+```python
+from sparseml.transformers.utils import SparseAutoModelForCausalLM
+
+# should contain model.safetensors or model.safetensors.index.json
+model_path = "/PATH/TO/COMPRESSED_MODEL"
+
+model = SparseAutoModelForCausalLM.from_pretrained(
+    model_name_or_path=model_path,
+    **model_kwargs,
+)
+```
+
+Saving a compressed model with an explicitly provided compression config. The config
+is saved to the model's `config.json` file. **Note:** the model must have been 
+initialized with SparseAutoModelForCausalLM.from_pretrained()
+
+```python
+from sparseml.transformers.compression import BitmaskConfig
+
+output_dir = "/PATH/TO/SAVE/COMPRESSED_MODEL"
+sparsity_config = BitmaskConfig()
+
+model.save_pretrained(
+    save_directory=output_dir,
+    sparsity_config=sparsity_config,
+)
+```
+
+Saving a compressed model, inferring the config from the model attributes
+
+```python
+model.save_compressed(
+    model,
+    save_directory=output_dir,
+)
+
+# alternative
+model.save_pretrained(
+    save_directory=output_dir,
+    save_compressed=True
+)
+```
+
+Saving a model in the dense format, but still include a sparsity config in `config.json`
```

Comment:
Could this just be the default if the user does not provide `save_compressed=True`?

I am thinking of the following scenario:

- User prunes model in SparseML
- User exports model with `model.save_pretrained(save_directory=output_dir)`, forgetting to add a SparsityConfig
- User tries to load the model to use the sparse kernels in vllm. There is no sparsity config so vllm does not detect sparsity and it runs the dense kernels :(

I think this is very likely to happen in the case of new users.

In some regards, Im suggesting that every model exported from `SparseAutoModelForCausalLM` should have a `sparsity_config` associated with it

I think this will prevent a lot of issues down the road

An alternative would be to make `save_compressed=True` the default

Reply:
I did originally have the default as this DenseConfig(), but as @mgoin pointed out it requires we read the whole model to calculate the global sparsity which can take a bit of time for large models. For 7b models its not that bad, I think its takes about a minute, but could become unruly for larger models

I don't love the idea of compressing by default, because it limits users to only being able to load the model within our ecosystem. I think we either accept the overhead of calculating global sparsity, or we could have the default be to set the dense sparsity config without filling in all the details

Reply:
That makes sense on why compressing by default is an issue. Agree especially for flows where someone fine-tunes after exporting or needs to evaluate

I think it is worth taking the slow export times to ensure that every model exported from SparseML can be loaded properly into nm-vllm. Thoughts @mgoin ?

Reply:
Yeah I'm glad you brought up the detection of sparsity in the model being important for algo selection, not just compression. I agree with your perspective and think we should calculate+store the sparsity in the config when using `SparseAutoModelForCausalLM`. Maybe a simple tqdm while calculating the sparsity would help the user know what's going on for a few minutes - it could also print beforehand that it is happening and note "it can be skipped by passing in X". Sorry for the back-and-forth @Satrat !